DAOS-16501 build: Support of mercury with ASan

knard38 opened the pull request:
### Description

When DAOS is built with sanitizers (`SANITIZERS=address,thread,undefined`), third-party
dependencies should also be instrumented to allow full end-to-end sanitizer coverage and
avoid false positives at ABI boundaries.

Previously, Mercury (the networking library used by DAOS) was always built with CMake's
`RelWithDebInfo` build type, regardless of the active sanitizer. Mercury's CMake build
system natively supports dedicated sanitizer build types (`Asan`, `Tsan`, `Ubsan`) that
configure the appropriate compiler flags.

This patch:
- **Detects the active sanitizer** from the SCons `SANITIZERS` variable during the Mercury
   build configuration step.
- **Selects the matching CMake build type** (`Asan`, `Tsan`, or `Ubsan`) when a sanitizer
   is active, falling back to `RelWithDebInfo` otherwise.
- **Adds `-DMERCURY_LIB_DEBUG_NAME_IS_RELEASE:BOOL=ON`** when a sanitizer build type is
   used, ensuring consistent library naming expected by DAOS's build system.

Additional minor fixes:
- Capitalizes "Google Sanitizers" in the SCons option description.
- Fixes a minor YAML example typo (`- ..` → `- ...`) in the developer documentation.
- Adds `--verbose` to the `fpm` packaging call for better build diagnostics.

### Steps for the author:

* [x] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [x] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [x] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [x] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Remove useless mercury patch

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Remove useless comma

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Remove useless codespell ignore

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Instrument mercury with libasan according to build option SANITIZERS

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Fix merge issues.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

daosbuild3 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16917/2/testReport/

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

jolivier23 reviewed (changes requested):
You need to add an entry to daos.spec changelog, bump the daos and mercury versions.  See this:

Without updating the mercury version, this will never build in CI because there is already a mercury to use.

https://github.com/daos-stack/daos/blob/master/docs/dev/development.md#updating-a-3rd-party-component

soumagne reviewed (changes requested):
ok based on Jeff's comment, I'll ask you to hold on this PR for now until I update mercury myself with this patch included. (There is also not much that you gain currently from building mercury with asan options, except slightly better traces)

soumagne commented:
@knard38 see new option I have added in https://github.com/mercury-hpc/mercury/commit/74dda4a6a161fceb3445d43ea4a3d71f3bd0cf79

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Integrate reviewers comments:
- Fix CI integration
- Use of official mercury patch 74dda4a

Miscellaneous fixes related to package build.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

knard38 commented:
> @knard38 see new option I have added in [mercury-hpc/mercury@74dda4a](https://github.com/mercury-hpc/mercury/commit/74dda4a6a161fceb3445d43ea4a3d71f3bd0cf79)

- [x] Use official mercury patch 74dda4a

knard38 commented:
> > You need to add an entry to daos.spec changelog, bump the daos and mercury versions. See this:
> > Without updating the mercury version, this will never build in CI because there is already a mercury to use.
> > https://github.com/daos-stack/daos/blob/master/docs/dev/development.md#updating-a-3rd-party-component
> 
> * [x]  Fix CI integration

Fixed with commit 5cb9bbeb6

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

jolivier23 commented on utils/rpms/daos.spec:
```diff
@@ -25,7 +25,7 @@
 Version:       2.7.101
-Release:       16%{?relval}%{?dist}
+Release:       17%{?relval}%{?dist}
 Summary:       DAOS Storage Engine
```
Because the daos changelog is now the only documentation we have that something changed.  I'm open to other options though.  Or maybe the gitlog on `utils/rpms/package_info.sh` is enough?

knard38 replied:
- [ ] Remove daos.spec update

knard38 replied:
> * [x]  Remove daos.spec update

Fixed with commit 8eb204a32

knard38 replied:
> For the Mercury change log update, I will wait for #17063 to be merged. Is that OK for you @soumagne and @jolivier23 ?

- [x] Restore Mercury changelog

knard38 replied:
> > For the Mercury change log update, I will wait for #17063 to be merged. Is that OK for you @soumagne and @jolivier23 ?
> 
> * [x]  Restore Mercury changelog

Fixed with commit 89f602f8f and 967bc9f9d

jolivier23 commented on debian/changelog:
```diff
@@ -1,3 +1,34 @@
+  [ Jeff Olivier]
+  * Make daos-spdk conflict with spdk
+
+ -- Jeff Olivier <jeffolivier@google.com>  Mon, 19 May 2025 14:12:00 -0700
```
The whole debian folder is no longer used since we produce .deb files using fpm.   We shoudl just remove the folder rather than updating this file

jolivier23 commented:
I pushed this: https://github.com/daos-stack/daos/pull/17050

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Integrate reviewers comments:
- Remove daos.spec update
- Remove Debian packaging update

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

jolivier23 commented:
@soumagne this seems like a better option for the change logs

https://github.com/daos-stack/daos/pull/17063#pullrequestreview-3404941845

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Integrate reviewers comments:
- Restore Mercury Changelog

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

Fix mercury release number

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: Support of mercury with ASan

- Fix invalid date in mercury changelog
- add verbose option to fpm to have meaningull error messages

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

knard38 commented:
@soumagne , @jolivier23 and @liw please could you have a look at this PR that I just merged with master.

kanard38 committed:
DAOS-16501 build: Fix CMake option ordering in Mercury sanitizer build

CMake requires all -D options to precede the source directory argument
(cmake [options] <path-to-source>). The CMAKE_BUILD_TYPE and
MERCURY_LIB_DEBUG_NAME_IS_RELEASE flags were appended after '../mercury',
placing them after the source directory and risking them being silently
ignored by CMake.

Use insert(-1, ...) instead of append() so both flags are inserted
before '../mercury' in the cmake command list.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

kanard38 committed:
DAOS-16501 build: add ASAN dependency to FPM package configuration

Add libasan as a dependency of the base daos package in the FPM
build script (daos.sh), replacing the BuildRequires entries that
were previously defined in the now-unused daos.spec file.

Distro-specific package names:
- RHEL/el: libasan
- SUSE: libasan8
- Ubuntu 24.04 / Debian: libasan8

The distro-specific package name is resolved via set_lib_name in
package_info.sh, following the existing convention.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

knard38 commented:
Apologies for the last force push: commit message was incorrect.

kanard38 committed:
DAOS-16501 build: fix invalid libasan pkg dependency

Move libasan dependency from the daos base package to daos-devel,
as it is only needed when building custom ASAN-instrumented RPMs
for support tasks.

Signed-off-by: Cedric Koch-Hofer <cedric.koch-hofer@hpe.com>

daosbuild3 commented:
Test stage Unit Test completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16917/22/testReport/

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

knard38 commented:
@soumagne , @jolivier23 and @liw please could you have a look at this PR that I just merged with master.

soumagne commented on utils/rpms/package_info.sh:
```diff
@@ -132,6 +132,9 @@ export uuid_lib
 
+set_lib_name libasan lib libasan libasan8 libasan8
+export libasan_lib
+
```
are we building rpms with asan ?

knard38 replied:
No, the DAOS RPMs themselves are not built with ASan. This PR adds support so that when DAOS is built **from source** with `SANITIZERS=address` (or `thread`/`undefined`), Mercury is also instrumented with the matching sanitizer.

The `libasan` dependency is added to `daos-dev` to ensure the ASan runtime library is present on the system when a developer builds DAOS with ASan enabled. Without it, the ASan-instrumented Mercury would fail to link or run correctly.

We cannot use RPM `BuildRequires` for this because DAOS packaging uses `fpm`, which must produce both RPM and Debian packages — RPM-specific build dependencies would break Debian compatibility. Adding `libasan` as a dependency of `daos-dev` is the portable approach that works across both packaging formats.

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-16917/26/testReport/

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

kanard38 committed:
Merge remote-tracking branch 'origin/master' into ckochhof/dev/master/daos-16501-part1

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16917/29/execution/node/1265/log

knard38 commented:
@daos-stack/daos-gatekeeper please could you lend this PR with the following message:
- **title**:`DAOS-16501 build: Support of mercury with ASan`
- **body**:
```
Instrument mercury with libasan according to build option SANITIZERS
```
